homeassistant/helpers/entity_registry.py

```diff
@@ -616,6 +617,50 @@ def async_update_entity_options(
 
         return new
 
+    @callback
+    def async_migrate_entity_to_new_platform(
+        self,
+        entity_id: str,
+        new_platform: str,
+        *,
+        new_unique_id: str | None = None,
+        new_config_entry: ConfigEntry | None = None,
+    ) -> RegistryEntry | None:
+        """
+        Migrate entity to new platform.
+
+        This should only be used when an entity needs to be migrated between
+        integrations.
+        """
+        if not (entry := self.async_get(entity_id)):
+            return None
+        if (
+            state := self.hass.states.get(entity_id)
+        ) is not None and state.state != STATE_UNKNOWN:
+            raise ValueError("Only entities that haven't been loaded can be migrated")
+        self.async_remove(entity_id)
+        # Precreate the entity so that any customizations can be preserved
+        new_entry = self.async_get_or_create(
```

Comment:
This is not correct because the new entity registry entry will have a new id.

Reply:
what would be referencing the old ID if the entity hasn't been loaded yet?

Reply:
Instead of adding all this logic, what if we just simply evolve the RegistryEntry with the new platform and call it a day? If the integration needs to update anything else on the entity, it can do so via `async_update_entity` and we don't then have to duplicate any logic here.

Reply:
OK, so I've updated the logic. We still accept things that might change as part of a change between platforms but now we keep the same entry ID